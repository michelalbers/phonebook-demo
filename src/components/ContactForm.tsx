import { Button, Grid, TextField } from "@mui/material";
import { FormikErrors, useFormik } from "formik";
import React from "react";
import { Contact } from "../hooks/useContactDb";

export const ContactForm: React.FC<{
  contact?: Contact;
  onSave: (values: Omit<Contact, "id">) => void;
}> = (props) => {
  const form = useFormik({
    initialValues: props.contact
      ? {
          name: props.contact.name,
          phoneNumber: props.contact.phoneNumber,
        }
      : {
          name: "",
          phoneNumber: "",
        },
    validate: (values) => {
      const errors: FormikErrors<typeof values> = {};

      if (!values.name || values.name.length < 3)
        errors.name = "Please provide a name with at least 3 characters";

      if (!values.phoneNumber || !values.phoneNumber.match(/^\+[0-9]{5,}$/))
        errors.phoneNumber =
          "Please provide a phonenumber with at least 5 digits";

      return errors;
    },
    onSubmit: (values) => {
      props.onSave(values);
    },
  });

  return (
    <form onSubmit={form.handleSubmit}>
      <Grid container spacing={2}>
        <Grid item xs={12}>
          <TextField
            onChange={form.handleChange}
            InputLabelProps={{ shrink: true }}
            inputProps={{ "data-testid": "name-input" }}
            name="name"
            value={form.values.name}
            label="name"
            fullWidth
            helperText={form.touched.name && form.errors.name}
            error={form.touched.name && !!form.errors.name}
            placeholder="John Doe"
          />
        </Grid>
        <Grid item xs={12}>
          <TextField
            onChange={form.handleChange}
            InputLabelProps={{ shrink: true }}
            inputProps={{ "data-testid": "phonenumber-input" }}
            name="phoneNumber"
            value={form.values.phoneNumber}
            label="phone number"
            type="tel"
            fullWidth
            helperText={form.touched.phoneNumber && form.errors.phoneNumber}
            error={form.touched.phoneNumber && !!form.errors.phoneNumber}
            placeholder="+49123456789"
          />
        </Grid>
        <Grid item xs={12}>
          <Button
            disabled={!form.isValid}
            type={"submit"}
            fullWidth
            variant="contained"
            color="primary"
          >
            Save contact
          </Button>
        </Grid>
        <Grid item xs={12}>
          <Button fullWidth variant="text" href="/">
            Cancel
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};
