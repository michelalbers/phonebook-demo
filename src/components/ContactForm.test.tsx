import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import React from "react";
import { ContactForm } from "./ContactForm";

test("renders without errors and matches snapshot", () => {
  const tree = render(<ContactForm onSave={() => {}} />);
  expect(tree).toMatchSnapshot();
});

test("should display error messages when submitting an invalid form", async () => {
  render(<ContactForm onSave={() => {}} />);
  const submitButton = screen.getByText(/save contact/i);
  submitButton.click();

  await waitFor(() =>
    expect(screen.getByText(/please provide a name/gi)).toBeInTheDocument()
  );

  await waitFor(() =>
    expect(
      screen.getByText(/please provide a phonenumber/gi)
    ).toBeInTheDocument()
  );
});

test("should call onSave if submitting a valid form", async () => {
  const onSave = jest.fn();
  render(<ContactForm onSave={onSave} />);
  const submitButton = screen.getByText(/save contact/i);

  const nameInput = screen.getByTestId("name-input");
  const phoneNumberInput = screen.getByTestId("phonenumber-input");

  act(() => {
    fireEvent.change(nameInput, {
      target: { value: "Foo Name" },
    });

    fireEvent.change(phoneNumberInput, {
      target: { value: "+491234566789" },
    });
  });

  submitButton.click();

  await waitFor(() =>
    expect(onSave).toHaveBeenCalledWith({
      name: "Foo Name",
      phoneNumber: "+491234566789",
    })
  );
});

// TODO: Test rendering of an initial contact for edit mode
