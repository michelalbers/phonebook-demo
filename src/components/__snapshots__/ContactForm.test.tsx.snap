// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without errors and matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="mui-1"
                id="mui-1-label"
              >
                name
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="name-input"
                  id="mui-1"
                  name="name"
                  placeholder="John Doe"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1z7n62"
                  >
                    <span>
                      name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="mui-2"
                id="mui-2-label"
              >
                phone number
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="phonenumber-input"
                  id="mui-2"
                  name="phoneNumber"
                  placeholder="+49123456789"
                  type="tel"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1z7n62"
                  >
                    <span>
                      phone number
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Save contact
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <a
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-10vjzfb-MuiButtonBase-root-MuiButton-root"
              href="/"
              tabindex="0"
            >
              Cancel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="mui-1"
              id="mui-1-label"
            >
              name
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="name-input"
                id="mui-1"
                name="name"
                placeholder="John Doe"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1z7n62"
                >
                  <span>
                    name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="mui-2"
              id="mui-2-label"
            >
              phone number
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="phonenumber-input"
                id="mui-2"
                name="phoneNumber"
                placeholder="+49123456789"
                type="tel"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1z7n62"
                >
                  <span>
                    phone number
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Save contact
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-10vjzfb-MuiButtonBase-root-MuiButton-root"
            href="/"
            tabindex="0"
          >
            Cancel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
